src/parser/TripleComponent.h

```diff
@@ -215,54 +215,73 @@ class TripleComponent {
       const EncodedIriManager* encodedIriManager) const;
 
   // Convert the `TripleComponent` to an ID. If the `TripleComponent` is a
-  // string, the IDs are resolved using the `vocabulary`. If a string is not
-  // found in the vocabulary, `std::nullopt` is returned.
+  // Literal or IRI, the IDs are resolved using the `vocabulary`. If a Literal
+  // or IRI is not found in the vocabulary then its position in the vocabulary
+  // is returned.
   template <typename Vocabulary>
-  [[nodiscard]] std::optional<Id> toValueId(
-      const Vocabulary& vocabulary, const EncodedIriManager& evManager) const {
+  [[nodiscard]] std::variant<Id, std::pair<VocabIndex, VocabIndex>>
+  toValueIdOrBounds(const Vocabulary& vocabulary,
+                    const EncodedIriManager& evManager) const {
     AD_CONTRACT_CHECK(!isString());
     std::optional<Id> vid = toValueIdIfNotString(&evManager);
-    if (vid != std::nullopt) return vid;
+    if (vid != std::nullopt) return vid.value();
     AD_CORRECTNESS_CHECK(isLiteral() || isIri());
-    VocabIndex idx;
     const std::string& content = isLiteral()
                                      ? getLiteral().toStringRepresentation()
                                      : getIri().toStringRepresentation();
-    if (vocabulary.getId(content, &idx)) {
-      return Id::makeFromVocabIndex(idx);
+    auto [lower, upper] = vocabulary.getPositionOfWord(content);
+    if (lower != upper) {
+      return Id::makeFromVocabIndex(lower);
+    }
+    return std::make_pair(lower, upper);
```

Comment:
I believe by now every compiler should just support

```suggestion
    return std::pair(lower, upper);
```

with correct template deduction